packages/transaction-controller/src/hooks/SequentialPublishBatchHook.ts

```diff
@@ -0,0 +1,227 @@
+      }
+    }
+    log('Sequential publish batch hook completed', { results });
+    pendingTransactionTracker.stop();
+
+    return { results };
+  }
+
+  /**
+   * Waits for a transaction event (confirmed, failed, or dropped) and resolves/rejects accordingly.
+   *
+   * @param pendingTransactionTracker - The tracker instance to subscribe to events.
+   * @param transactionMeta - The transaction metadata.
+   * @param transactionHash - The hash of the transaction.
+   * @returns A promise that resolves when the transaction is confirmed or rejects if it fails or is dropped.
+   */
+  async #waitForTransactionEvent(
+    pendingTransactionTracker: PendingTransactionTracker,
+    transactionMeta: TransactionMeta,
+    transactionHash: string,
+  ): Promise<TransactionMeta> {
+    return new Promise((resolve, reject) => {
+      const onConfirmed = this.#onConfirmedHandler({
+        transactionMeta,
+        transactionHash,
+        resolve,
+        pendingTransactionTracker,
+      });
+
+      const onFailedOrDropped = this.#onFailedOrDroppedHandler({
+        transactionMeta,
+        transactionHash,
+        reject,
+        pendingTransactionTracker,
+      });
+
+      this.#addListeners({
+        pendingTransactionTracker,
+        onConfirmed,
+        onFailedOrDropped,
+      });
+    });
+  }
+
+  #onConfirmedHandler({
+    transactionMeta,
+    transactionHash,
+    resolve,
+    pendingTransactionTracker,
+  }: OnConfirmedHandlerOptions): (txMeta: TransactionMeta) => void {
+    return (txMeta) => {
+      if (txMeta.id === transactionMeta.id) {
+        log('Transaction confirmed', { transactionHash });
+        this.#removeListeners({
+          pendingTransactionTracker,
+        });
+        resolve(txMeta);
+      }
+    };
+  }
+
+  #onFailedOrDroppedHandler({
+    transactionMeta,
+    transactionHash,
+    reject,
+    pendingTransactionTracker,
+  }: OnFailedOrDroppedHandlerOptions): (
+    txMeta: TransactionMeta,
+    error?: Error,
+  ) => void {
+    return (txMeta, error) => {
+      if (txMeta.id === transactionMeta.id) {
+        log('Transaction failed or dropped', { transactionHash, error });
+        this.#removeListeners({
+          pendingTransactionTracker,
+        });
+        reject(new Error(`Transaction ${transactionHash} failed or dropped.`));
+      }
+    };
+  }
+
+  #addListeners({
+    pendingTransactionTracker,
+    onConfirmed,
+    onFailedOrDropped,
+  }: TrackerListenersOptions): void {
+    pendingTransactionTracker.hub.on('transaction-confirmed', onConfirmed);
+    pendingTransactionTracker.hub.on('transaction-dropped', onFailedOrDropped);
+    pendingTransactionTracker.hub.on('transaction-failed', onFailedOrDropped);
+  }
+
+  #removeListeners({
+    pendingTransactionTracker,
+  }: {
+    pendingTransactionTracker: PendingTransactionTracker;
+  }): void {
+    pendingTransactionTracker.hub.removeAllListeners('transaction-confirmed');
```

Comment:
How do we know that the controller itself or other components haven't registered any listeners? Should we store references to our specific ones so we can remove them?

Reply:
Now I'm saving the reference and calling `off` instead of removing all listeners.